### PR TITLE
Internal API Refactor

### PR DESCRIPTION
**Pull request by calebcase**
This is a complete rewrite of our internal API.

* Dropped support for 2.6 and 3.2. Neither of these are supported upstream anymore.
* Added support for 3.5.
* Using tox-travis now allows us to accurately test against specific versions using travis while retaining the flexibility of tox at home.
* A newly minted segments API:
  * Python 2 vs 3 agnostic.
  * Keys are expected to be tuples, but are allowed to be any sequence-like thing.
  * Generalized (key, value) generator that works for sequence-like and dictionary-like things.
  * Uses generators whenever possible.
  * Pythonic duck typing is employed to make decisions between sequences-like and dictionary-like things.
  * walk, fold, foldm are very handy.
  * Thorough testing accomplished through the hypothesis framework. Fuzzing our way to happiness.
* Fixes some gnarly bugs. Yay!
* Version bumped to 1.5

### CODE REVIEW

File: dpath/segments.py [hunk 1]
```diff
@@ -0,0 +1,386 @@
...
+    except TypeError:
+        # We attempted to extend something that doesn't support it. In
+        # this case we assume thing is actually more like a dictionary
+        # and doesn't need to be extended.
+        pass
+
+    return thing
+
+
+def set(obj, segments, value, create_missing=True, hints=()):
```
Comment (calebcase) [inline dpath/segments.py]: Need to add hints to docstring.

File: dpath/util.py [hunk 2]
```diff
@@ -172,48 +232,74 @@ def merge(dst, src, separator="/", afilter=None, flags=MERGE_ADDITIVE, _path="")
...
-                dst += [None] * (dsti - (len(dst) - 1))
-            if dst[dsti] == None:
-                dst[dsti] = src[i]
-            else:
-                if not isinstance(src[i], (MutableMapping, MutableSequence)):
-                    dst[dsti] = src[i]
-                else:
-                    merge(dst[i], src[i], afilter=afilter, flags=flags,
-                          _path=separator.join([_path, str(i)]), separator=separator)
+            if len(key) == 0 and not options.ALLOW_EMPTY_STRING_KEYS:
```
Comment (calebcase) [inline dpath/util.py]: Will fail on integer keys, need to use the same try: lenght = len(key) trick used in walk().

File: tests/test_path_get.py [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import nose
```
Comment (calebcase) [inline tests/test_path_get.py]: Need to rename to test_segments_get.py

File: tests/test_path_paths.py [hunk 4]
```diff
@@ -1,6 +1,6 @@
 import nose
```
Comment (calebcase) [inline tests/test_path_paths.py]: Need to rename to test_segments_walk.py

File: tests/test_util_new.py [hunk 5]
```diff
@@ -29,7 +29,7 @@ def test_set_new_list():
     dpath.util.new(dict, '/a/1', 1)
     assert(dict['a'][1] == 1)
     assert(dict['a'][0] == None)
-    dpath.util.new(dict, ['a', '1'], 1)
+    dpath.util.new(dict, ['a', 1], 1)
```
Comment (calebcase) [inline tests/test_util_new.py]: Should revert this back. It should handle this now.

File: tests/test_util_search.py [hunk 6]
```diff
@@ -172,12 +172,12 @@ def test_search_return_list_globbed():
         }
     res = dpath.util.search(tdict, '/a/b/[02]')
     assert(isinstance(res['a']['b'], list))
-    assert(len(res['a']['b']) == 2)
-    assert(res['a']['b'] == [0, 2])
+    assert(len(res['a']['b']) == 3)
+    assert(res['a']['b'] == [0, None, 2])
```
Comment (calebcase) [inline tests/test_util_search.py]: Note: Change of behavior here, for the better I hope.

**Comment by calebcase** [comment]
If I can find the time, we really should give dpath/util.py the hypothesis treatment as well.

File: tests/test_util_search.py (hunk 6 above)
Comment (akesterson) [inline tests/test_util_search.py]: I had to re-read the docs here to make sure what this test is asserting, and I'm still not sure. At first I thought "is it intending to retrieve indexes 0 and 2", but then after reading the docs I have to assume it's matching any keys in the object that match the glob "[02]" - which appears to only happen because the keys get interpolated as strings (which is, I'm pretty sure, not our intention).

All that being said, I can't see any interpretation of util.search(x, '/a/b/[02]') where the user would expect the result to be [0, None, 2]. What's the rationale behind this change here?

File: dpath/segments.py [hunk 7]
```diff
@@ -0,0 +1,386 @@
...
+    If the segment is an integer, return the string conversion.
+    Otherwise return the segment unchanged. The conversion uses 'str'
+    which means that:
+
+        * in Python 2.x: 0 -> b'0'
+        * in Python 3.x: 0 -> u'0'
+
+    int_str(segment) -> str
+    '''
+    if isinstance(segment, (int, long)):
```
Comment (akesterson) [inline dpath/segments.py]: While it's incredibly unlikely that someone will subclass int/long, would it be worth it to change these to issubclass() calls (both here and throughout)? Our bugtracker is full of edge cases lately, and this one seems easy to head off at the pass.

Comment (calebcase) [inline dpath/segments.py]: Sounds solid to me.

File: tests/test_util_search.py (hunk 6 above)
Comment (calebcase) [inline tests/test_util_search.py]: `util.search(x, '/a/b/[02]')` => This to me means: Search through x for paths that match /a/b/0 and /a/b/2 and return a view of the result. Since x['a']['b'] is a sequence we have to fill in the missing elements with something or else the view is not self consistent (and doesn't satisfy the doc string requirement that it should contain the keys that matched the glob).

By self consistent I mean that if I do a search I should be able to also do a search on that search and get the same result:

```
v1 = util.search(x, '/a/b/[02]')
v2 = util.search(v1, '/a/b/[02]')
assert(v1 == v2)
```

It doesn't make sense to me to have a search that returns a resulting view that then wouldn't match the original search.

It may be that we need either a different function or a flag to control the behavior when creating sparse arrays.

**Comment by calebcase** [comment]
It also occurs to me that the walk termination is now driven by specific leaf types vs the previous implementation being driven by not being a MutableSequence/MutableDictionary. I think we may be looking at a major version bump instead of a minor. In which case we may want to RC this and do some of the other API cleanup to support various behaviors in walk, leaves, sparse arrays, etc.

**Comment by akesterson** [comment]
After some discussion, here's some of the challenges presented here:

* In order to truly "solve" the unicode problem, we had to engage in the grand majority of the rewrite seen in the segments module.
* We can't easily make the new segments bit work in a backwards compatible fashion with respect to some of the list handling (see the [0, None, 2] discussion)
* We want to take this refactor further and make it more extensible for a wider variety of user use cases
* We can't do the refactor when everything is still tied up in the segments module, and the user API remains unchanged

.. To address this, we've decided to merge this request into a branch for a complete 2.0 refactor of dpath. We'll be centering the refactor around the new work in segments, but this will be broken out into some kind of behavior class which can be overriden by users and passed around, keeping the API pretty clean, but without making users instantiate our API directly every single time.

I have merged this PR to the 2.0 branch. Marking this request as closed. Further work will be discussed in the README/Wiki and pull requests for the 2.0 branch.

I will look into whether or not the few existing bugs can be closed without this refactor. Some of them will probably have to wait for 2.0.
